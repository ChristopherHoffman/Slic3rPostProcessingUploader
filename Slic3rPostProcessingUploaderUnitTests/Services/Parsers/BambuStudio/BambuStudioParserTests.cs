using Slic3rPostProcessingUploader.Services.Parsers.AnycubicSlicerNext;
using Slic3rPostProcessingUploader.Services.Parsers.BambuStudio;
using Slic3rPostProcessingUploaderUnitTests.Services.Parsers.AnycubicSlicerNext;
using Snapshooter.MSTest;


namespace Slic3rPostProcessingUploaderUnitTests.Services.Parsers.BambuStudio
{
    [TestClass]
    public class BambuStudioParserTests
    {
        [TestMethod]
        public void ShouldReturnAnEmptySettingWhenGcodeIsEmpty()
        {
            var parser = new BambuStudioParser("");
            var result = parser.ParseGcode("");
            Assert.IsNotNull(result);

            Snapshot.Match(result);
        }

        [TestMethod]
        public void ShouldReturnExpectedValuesWhenGivenFullGcode()
        {
            var parser = new BambuStudioParser("");
            var result = parser.ParseGcode(BambuStudioParserTestGcode.CalibrationCube);

            Snapshot.Match(result);
        }

        [TestMethod]
        public void ShouldRenderTheExpectedNoteWhenGivenATemplateWithNoReplacements()
        {
            string template = "Settings:";

            var parser = new BambuStudioParser(template);
            var result = parser.ParseGcode(BambuStudioParserTestGcode.CalibrationCube);

            Assert.AreEqual("Settings:", result.settings.note);
        }

        [TestMethod]
        public void ShouldRenderTheExpectedNoteWhenGivenATemplateWithASingleReplacement()
        {
            string template = """
                Settings:
                    Layer Height: {{layer_height}}
                """;

            var parser = new BambuStudioParser(template);
            var result = parser.ParseGcode(BambuStudioParserTestGcode.CalibrationCube);

            Assert.AreEqual("""
                Settings:
                    Layer Height: 0.2
                """, result.settings.note);
        }

        [TestMethod]
        public void ShouldRenderTheExpectedNoteWhenGivenATemplateWithMultipleReplacements()
        {
            string template = """
                Settings:
                    Layer Height: {{layer_height}}
                    First Layer Height: {{initial_layer_print_height}}
                    Wall Loops: {{wall_loops}}
                    Top Shell Layers: {{top_shell_layers}}
                    Bottom Shell Layers: {{bottom_shell_layers}}
                    Sparse Infill Density: {{sparse_infill_density}}
                """;

            var parser = new BambuStudioParser(template);
            var result = parser.ParseGcode(BambuStudioParserTestGcode.CalibrationCube);

            Assert.AreEqual("""
                Settings:
                    Layer Height: 0.2
                    First Layer Height: 0.28
                    Wall Loops: 3
                    Top Shell Layers: 3
                    Bottom Shell Layers: 3
                    Sparse Infill Density: 10%
                """, result.settings.note);
        }

        [TestMethod]
        public void ShouldRenderFullTemplateWhenGivenAGcodeWithTwoFilaments()
        {
            var parser = new BambuStudioParser("");
            var result = parser.ParseGcode(BambuStudioParserTestGcode.CalibrationCubeTwoFilament);

            Snapshot.Match(result, matchOptions => matchOptions.HashField("settings.Snapshot"));
        }

    }
}
