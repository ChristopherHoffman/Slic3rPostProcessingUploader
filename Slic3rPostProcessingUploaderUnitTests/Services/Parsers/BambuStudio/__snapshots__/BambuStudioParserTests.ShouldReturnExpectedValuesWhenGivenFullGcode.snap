{
  "Slicer": "BambuStudioSlicer",
  "CuraVersion": "Unknown",
  "PluginVersion": null,
  "settings": {
    "note": "Settings:\n\nQuality:\n  Layer Height:\n    Layer Height: 0.2\n    First Layer Height: 0.28\n  Line Width:\n    Default Line Width: 0.42\n  Seam:\n    Seam Position: aligned\n  Precision:\n    Precise Wall: \n  Ironing:\n    Ironing: no ironing\n  Wall Generator:\n    Generator: arachne\n  Walls And Surfaces:\n    Only One Wall on Top Surfaces: \n\nStrength:\n  Walls:\n    Wall Loops: 3\n  Top/Bottom Shells:\n    Top Shell Layers: 3\n    Top Shell Thickness: 0.6\n    Bottom Shell Layers: 3\n    Bottom Shell Thickness: 1.2\n  Infill:\n    Sparse Infill Density: 10%\n    Sparse Infill Pattern: zig-zag\n\nSpeed:\n  First Layer Speed:\n    First Layer: 45\n    Number of Slow Layers: \n  Other Layers Speed:\n    Outer Wall: 150\n    Inner Wall: 150\n    Sparse Infill: 70\n    Internal Solid Infill: 150\n    Top Surface: 120\n  Travel Speed:\n    Travel: 120\n\nSupport:\n  Support:\n    Enable Support: 0\n    Type: normal(auto)\n    Style: default\n    Threshold Angle: 30\n    On Build Plate Only: 0\n  Raft:\n    Raft Layers: 0\n\nMultimaterial:\n  Prime Tower:\n    Enable: 0\n    Width: 60\n    Brim Width: 3\n  Ooze Prevention:\n    Enable: 0\n  Flush Options:\n    Flush into Objects' Infill: 0\n    Flush into Objects' Support: 1\n  Advanced:\n    Use Beam Interlocking: \n\nOther:\n  Skirt:\n    Skirt Type: \n    Skirt Loops: 0\n    Skirt Min Extrusion Length: \n  Brim:\n    Brim Type: auto_brim\n    Brim Width: 3\n  Special Mode:\n    Slicing Mode: regular\n    Print Sequence: by layer\n    Spiral Vase: 0\n    Fuzzy Skin: none",
    "print_name": null,
    "time_stamp": null,
    "estimated_print_time_seconds": 996,
    "material_used_mg": 4440,
    "Snapshot": "iVBORw0KGgoAAAANSUhEUgAAADIAAAAyCAYAAAAeP4ixAAADGklEQVR4Ae2Y204TURSG1wWJeCCS2q\nSJGKgtFloQag8cCk24MNx40+iNKFE0TWNUsJ6i0Wg9XFTroRoNUaOBN+gj8Ah9hD5CH2Hsz3bbod3T\nzkxnOjPJ/MkKhNmF9fGv9e+mRK5cubJWRwaGyTvoJ0dr8vgKbS/W9yt2IkOO1JVAmXbT0oHCz+CQIz\nR6NEpvYtU2CF54hjO21upIXhFAXhg1nLWdO2jo6cyeKgh54TW2CYJl34awye8LEn1bUOcOfoelyoZ2\nhM0VohJdG5foZVS9O1uRSv9HDcv6MVlra+b3kkS3wwwCMK/PMWfUwuB39i2mM6MFYRPFBAMACG/+1x\nKD0bI3GDX8DdPc6RSrT84yiOez7c8AAiAtMDwIDI/pTrGKccIYwRGlpS/GtYPIY9oQZcbKupqQF5b+\nz7L42c9U99fnJirUs9aDVSollBtRU1/n212Bkw+nJbrb2Kkv88qvhdsXx2rUs9ZOV2mz8ceezTRmPS\nXRblpfyV0pNxq/GWL1ISk+/yPVTEBDQVD5iLTvjh6Qz3OsEAhobivCXBGd5QnIy3AQXnBH66jh/L0I\na+xtTPkcT8C+gHB3Os22qN4n2MiIniHdchPtEKaD8MJ/t5cgQPFxU6q+gKAwDtuL2gHgAl9oXtmQhS\nDyIFDrDpxsbfjxNCtLQeRB0Cmm+f3R2uyNMxK9i1s8WiJ3PgnuB35/iJq9P8XuGluB8HoluwS7LTQu\nxmzIpiA8CESxejUo0SU/+/5OWNkN24BshpUbXA/iIyIGcWvSwSB8yUVJ5UgQOLMWYF8dCQKIUrIJg7\nMiGENA5rwF00AeNUbqxb8lL0Sb75JbzyW9ZTJE3sEo5UJ1w0FKssjF/QKY1oTzHV4hw3X+5I5hIHiP\nVYwrQ66O7JGpCgxlNLsjahQjhdtc9Cw8bNAHDt2EUbtwqqIbBMst2gW44DlkwSf1aV9ZFwhu+wdTJi\n20XsGd6+M1TSDyN46XA3VzFlqvOgWBZQutV0pBIIIIDm2QrYVRa3Wn1YVjA35yjKKe/H93OMSsp0CO\nFA8C2y20K1euHKu/gG1NvibXFfQAAAAASUVORK5CYII=",
    "file_name": null
  }
}
